drivers/rtc/rtc_ifx_cat1.c

```diff
@@ -0,0 +1,338 @@
+ */
+#define PPB_TO_WCO_PULSE_SETS(ppb) ((ROUND_DOWN((ppb * 9216 / 78125), 128)) / 128)
+
+/* Convert groupings of 128 ticks added or removed from one hour of clock cycles at
+ * 32768 Hz to parts per billion
+ *
+ * wps * 128 * 1000000000 / 32768Hz * 60min * 60sec
+ * wps * 128000000000 / 117964800
+ * wps * 78125 / 72
+ */
+#define WCO_PULSE_SETS_TO_PPB(wps) (wps * 78125 / 72)
+
+struct ifx_cat1_rtc_config {
+	uint32_t irqn;
+};
+
+struct ifx_cat1_rtc_data {
+	struct k_mutex lock;
+};
+
+static inline uint16_t _ifx_cat1_rtc_get_state(void)
+{
+	return _FLD2VAL(_IFX_CAT1_RTC_BREG_STATE, _IFX_CAT1_RTC_BREG);
+}
+
+static inline void _ifx_cat1_rtc_set_state(uint16_t init)
+{
+	_IFX_CAT1_RTC_BREG &= _IFX_CAT1_RTC_BREG_CENTURY_Msk;
+	_IFX_CAT1_RTC_BREG |= _VAL2FLD(_IFX_CAT1_RTC_BREG_STATE, init);
+}
+
+static inline uint16_t _ifx_cat1_rtc_get_century(void)
+{
+	return _FLD2VAL(_IFX_CAT1_RTC_BREG_CENTURY, _IFX_CAT1_RTC_BREG);
+}
+
+static inline void _ifx_cat1_rtc_set_century(uint16_t century)
+{
+	_IFX_CAT1_RTC_BREG &= _IFX_CAT1_RTC_BREG_STATE_Msk;
+	_IFX_CAT1_RTC_BREG |= _VAL2FLD(_IFX_CAT1_RTC_BREG_CENTURY, century);
+}
+
+static void _ifx_cat1_rtc_from_pdl_time(cy_stc_rtc_config_t *pdlTime, const int year,
+					struct rtc_time *z_time)
+{
+	CY_ASSERT(pdlTime != NULL);
+	CY_ASSERT(z_time != NULL);
+
+	/* The number of days that precede each month of the year, not including Feb 29 */
+	static const uint16_t CUMULATIVE_DAYS[] = {0,   31,  59,  90,  120, 151,
+						   181, 212, 243, 273, 304, 334};
+	z_time->tm_sec = (int)pdlTime->sec;
+	z_time->tm_min = (int)pdlTime->min;
+	z_time->tm_hour = (int)pdlTime->hour;
+	z_time->tm_mday = (int)pdlTime->date;
+	z_time->tm_mon = (int)(pdlTime->month - 1u);
+	z_time->tm_year = (int)(year - _IFX_CAT1_RTC_TM_YEAR_BASE);
+	z_time->tm_wday = (int)(pdlTime->dayOfWeek - 1u);
+	z_time->tm_yday = (int)CUMULATIVE_DAYS[z_time->tm_mon] + (int)pdlTime->date - 1 +
+			  (((int)(pdlTime->month) >= 3 &&
+			    (int)(Cy_RTC_IsLeapYear((uint32_t)year) ? 1u : 0u)));
+
+	/* daylight savings currently marked as unknown */
+	z_time->tm_isdst = -1;
+
+	/* nanoseconds not tracked by ifx code. Set to value indicating unknown */
+	z_time->tm_nsec = 0;
+}
+
+static void _ifx_cat1_rtc_isr_handler(void)
+{
+	Cy_RTC_Interrupt(_ifx_cat1_rtc_dst, NULL != _ifx_cat1_rtc_dst);
+}
+
+void _ifx_cat1_rtc_century_interrupt(void)
+{
+	_ifx_cat1_rtc_set_century(_ifx_cat1_rtc_get_century() + 100);
+}
+
+static int ifx_cat1_rtc_init(const struct device *dev)
+{
+	cy_rslt_t rslt = CY_RSLT_SUCCESS;
+	static const cy_stc_rtc_config_t default_time = {
+		.dayOfWeek = CY_RTC_SATURDAY,
+		.date = 1,
+		.month = 1,
+		.year = 0,
+	};
+
+	Cy_SysClk_ClkBakSetSource(CY_SYSCLK_BAK_IN_CLKLF);
+
+	if (_ifx_cat1_rtc_get_state() == _IFX_CAT1_RTC_STATE_UNINITIALIZED) {
+		if (Cy_RTC_IsExternalResetOccurred()) {
+			/* Reset to default time */
+			Cy_RTC_SetDateAndTime(&default_time);
+			_ifx_cat1_rtc_set_century(_IFX_CAT1_RTC_INIT_CENTURY);
+		}
```

Comment:
The expected behavior of the RTC is to return `-ENODATA` until the time has actually been set to a valid value, rather than defaulting to a pseudo random value.

The init function should just do the minimal amount of init, like configuring the IRQ. Setting the time is done with `rtc_set_time()` at which point the RTC's time will be initialized to a valid value